### PR TITLE
RegisterProvider engine work

### PR DESCRIPTION
**Pull request by Frassle**
# Description

This adds support for a `RegisterProvider` method to the engine. This allows an SDK process to send the information for a package (name, version, url, etc, and parameter in the future) and get back a UUID for that run of the engine that can be used to re-lookup that information.

That allows the SDK to just send the `provider` field in `RegisterResourceRequest` instead of filling in `version`, `pluginDownloadURL` etc (and importantly not having to fill in `parameter` for parameterised providers, which could be a large amount of data).

This doesn't update any of the SDKs to yet use this method. We can do that piecemeal, but it will require core sdk and codegen changes for each language.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/engine/lifecycletest/parameterized_test.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
+// Copyright 2016-2022, Pulumi Corporation.
```
Comment (tgummerer) [inline pkg/engine/lifecycletest/parameterized_test.go:1]:
```suggestion
// Copyright 2024-2024, Pulumi Corporation.
```

File: pkg/resource/deploy/source_eval.go [hunk 2]
```diff
@@ -782,6 +797,41 @@ func parseProviderRequest(
...
+		v, err := semver.Parse(req.Version)
+		if err != nil {
+			return nil, fmt.Errorf("parse package version %s: %w", req.Version, err)
+		}
+		version = &v
+	}
+	pi := providers.NewProviderRequest(
+		version, tokens.Package(req.Name), req.PluginDownloadUrl, req.PluginChecksums)
+
+	rm.packageRefLock.Lock()
```
Comment (tgummerer) [inline pkg/resource/deploy/source_eval.go:817]: I was wondering if it was worth using a readlock here and then upgrading to a writelock later.  But iiuc this only needs to be called once per component, not once per resonrce, right?  So this should not be called very often per pulumi program, and should be okay to just have this one lock, and having a read lock is a bit more tricky anyway.

File: pkg/resource/deploy/source_query.go [hunk 3]
```diff
@@ -332,6 +333,50 @@ func (rm *queryResmon) Cancel() error {
 	return <-rm.done
 }
 
+// getProviderReference fetches the provider reference for a resource, read, or invoke from the given package with the
+// given unparsed provider reference. If the unparsed provider reference is empty, this function returns a reference
+// to the default provider for the indicated package.
+func getProviderReference(defaultProviders *defaultProviders, req providers.ProviderRequest,
```
Comment (tgummerer) [inline pkg/resource/deploy/source_query.go:339]: So we're not going to support parameterized providers in `pulumi query`, which makes sense to me.  Should we at return an error response with a little bit more information if `RegisterProvider` is called? I think the answer is no, and whatever error message Grpc gives us in this case is good enough, but figured I ask the question anyway.

File: proto/pulumi/resource.proto [hunk 4]
```diff
@@ -238,3 +240,15 @@ message TransformResponse {
     google.protobuf.Struct properties = 1; // the transformed input properties.
     TransformResourceOptions options = 2; // the options for the resource.
 }
+
+message RegisterProviderRequest {
```
Comment (tgummerer) [inline proto/pulumi/resource.proto:244]: I might be missing something here, but don't we need to send the parameters for parameterized providers here as well?  Or is that all encoded in the checksums?

File: pkg/resource/deploy/source_eval.go (hunk 2 above)
Comment (Frassle) [reply pkg/resource/deploy/source_eval.go:817]: yeh once per process generally

File: pkg/resource/deploy/source_query.go (hunk 3 above)
Comment (Frassle) [reply pkg/resource/deploy/source_query.go:339]: > So we're not going to support parameterized providers

No intention for it correct.

> I think the answer is no, and whatever error message Grpc gives us in this case is good enough,

Yeh just "Not Implemented" is fine. The SDK will catch and handle that, needs to handle it for old engine as well anyway.

File: proto/pulumi/resource.proto (hunk 4 above)
Comment (Frassle) [reply proto/pulumi/resource.proto:244]: We do need to send the parameters. But we haven't yet designed that fully yet.

**Review by tgummerer — APPROVED** [review]

**Pull request merged**